numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -2493,9 +2574,12 @@ string_multiply_promoter(PyUFuncObject *ufunc_obj, PyArray_DTypeMeta *op_dtypes[
         if (signature[i]) {
             tmp = signature[i];
         }
-        else if (op_dtypes[i] != &PyArray_StringDType) {
+        else if (is_signed_integer_dtype(op_dtypes[i])) {
```

Comment:
I think I'm getting a little less confused, but wonder if this can still be easier: If we just register one promoter for `None` that just sets `int64` here and another one for `uint64` that sets `uint64`, would we then not be OK without the `is_(un)signed_integer_dtype` functions? That would provide safe casting routes for all integers, no?

Reply:
That can't work because I'm installing the promoter with a `None` entry in the dtypes I'm installing it for. That makes the promoter fire for all dtypes. There's no way to say "i only want to install a promoter for unsigned integers" without explicitly adding promoters for all unsigned integers like I had before.

Reply:
I meant that `None` would be for all integers except `uint64` -- which is the only one that cannot be cast to `int64`. I think that if the `uint64` is registered first, that should work, no? But I must admit that even though I reviewed @seberg's PRs, I fear I don't completely understand the paths yet...

Reply:
Registration order is meaningless, subclass relationships define priority. `None` is just the same as `np.dtype` (always matches any dtype with lowest priority).

But I think Marten is right (with caveat):  If you install the `None` promoter it is guaranteed to never trigger for `uint64` because there is a more specific `uint64` loop already registered.

The (big) caveat is that, unfortunately, `ulong` and `ulonglong` are both `unit64` and I never made them subclasses of each other :(.  And until they are, you need to register _both_ (that could be another promoter to convert the `ulonglong` to `uint64` or another loop)

Reply:
(I am wondering now if we can try to hack that subclass relationship without merging the two uint64 into a single DType, but not for the here and now!)

Reply:
Promoters don't have *any* performance implications at all since the result is always cached (a bit too much if anything).  The only question is what is nicer.

Reply:
Yes, the only reason I brought it up was that I thought it would be nicer without the `is_integer` function, since it just seemed promoters should be able to handle this - so a bit less code (both source and compiled). But the beauty is slightly destroyed by needing different ones for `ulong` and `ulonglong`...

To me, the errors do not make too much of a difference - in a way, it is nice that you can do `casting='unsafe'` for float. 

But really this is not a big deal - I really just hoped we could avoid registering a whole slew of loops and that is done now! So, @ngoldbaum, happy to go with what you feel is nicest here!